src/algebra/algebra/tower.lean

```diff
@@ -0,0 +1,281 @@
+  (λ x y z, add_smul x y z)
+  (λ c x y, smul_assoc c x y)
+  (λ x y z, smul_add x y z)
+  (λ c x y, smul_algebra_smul_comm c x y)) }
+
+@[simp] lemma lsmul_coe (a : A) : (lsmul R M a : M → M) = (•) a := rfl
+
+end algebra
+
+namespace is_scalar_tower
+
+section semimodule
+
+variables [comm_semiring R] [semiring A] [algebra R A]
+variables [add_comm_monoid M] [semimodule R M] [semimodule A M] [is_scalar_tower R A M]
+
+variables {R} (A) {M}
+theorem algebra_map_smul (r : R) (x : M) : algebra_map R A r • x = r • x :=
+by rw [algebra.algebra_map_eq_smul_one, smul_assoc, one_smul]
+
+variables {R A M}
+theorem smul_left_comm (r : R) (s : A) (x : M) : r • s • x = s • r • x :=
+by rw [← smul_assoc, algebra.smul_def r s, algebra.commutes, mul_smul, algebra_map_smul]
+
+
+end semimodule
+
+section semiring
+variables [comm_semiring R] [comm_semiring S] [semiring A] [semiring B]
+variables [algebra R S] [algebra S A] [algebra S B]
+
+variables {R S A}
+theorem of_algebra_map_eq [algebra R A]
+  (h : ∀ x, algebra_map R A x = algebra_map S A (algebra_map R S x)) :
+  is_scalar_tower R S A :=
+⟨λ x y z, by simp_rw [algebra.smul_def, ring_hom.map_mul, mul_assoc, h]⟩
+
+variables (R S A)
+
+instance subalgebra (S₀ : subalgebra R S) : is_scalar_tower S₀ S A :=
+of_algebra_map_eq $ λ x, rfl
+
+variables [algebra R A] [algebra R B]
+variables [is_scalar_tower R S A] [is_scalar_tower R S B]
+
+theorem algebra_map_eq :
+  algebra_map R A = (algebra_map S A).comp (algebra_map R S) :=
+ring_hom.ext $ λ x, by simp_rw [ring_hom.comp_apply, algebra.algebra_map_eq_smul_one,
+    smul_assoc, one_smul]
+
+theorem algebra_map_apply (x : R) : algebra_map R A x = algebra_map S A (algebra_map R S x) :=
+by rw [algebra_map_eq R S A, ring_hom.comp_apply]
+
+instance subalgebra' (S₀ : subalgebra R S) : is_scalar_tower R S₀ A :=
+@is_scalar_tower.of_algebra_map_eq R S₀ A _ _ _ _ _ _ $ λ _,
+(is_scalar_tower.algebra_map_apply R S A _ : _)
+
+@[ext] lemma algebra.ext {S : Type u} {A : Type v} [comm_semiring S] [semiring A]
+  (h1 h2 : algebra S A) (h : ∀ {r : S} {x : A}, (by haveI := h1; exact r • x) = r • x) : h1 = h2 :=
+begin
+  unfreezingI { cases h1 with f1 g1 h11 h12, cases h2 with f2 g2 h21 h22,
+  cases f1, cases f2, congr', { ext r x, exact h },
+  ext r, erw [← mul_one (g1 r), ← h12, ← mul_one (g2 r), ← h22, h], refl }
+end
+
+variables (R S A)
+theorem algebra_comap_eq : algebra.comap.algebra R S A = ‹_› :=
+algebra.ext _ _ $ λ x (z : A),
+calc  algebra_map R S x • z
+    = (x • 1 : S) • z : by rw algebra.algebra_map_eq_smul_one
+... = x • (1 : S) • z : by rw smul_assoc
+... = (by exact x • z : A) : by rw one_smul
+
+/-- In a tower, the canonical map from the middle element to the top element is an
+algebra homomorphism over the bottom element. -/
+def to_alg_hom : S →ₐ[R] A :=
+{ commutes' := λ _, (algebra_map_apply _ _ _ _).symm,
+  .. algebra_map S A }
+
+@[simp] lemma to_alg_hom_apply (y : S) : to_alg_hom R S A y = algebra_map S A y := rfl
```

Comment:
I think this might be better as `to_alg_hom_coe`, equating  coe_fn's

Reply:
Done.

Reply:
Although it might be worth keeping this as a non-simp lemma anyway

Reply:
I did.